### PR TITLE
Parsing of immutable state variable.

### PR DESCRIPTION
**Pull request by chriseth**
References #3835

### CODE REVIEW

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 1]
```diff
@@ -378,7 +380,7 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 	else if (_variable.isStateVariable())
 	{
 		solAssert(varLoc == Location::Unspecified, "");
-		typeLoc = _variable.isConstant() ? DataLocation::Memory : DataLocation::Storage;
+		typeLoc = (_variable.isConstant() || _variable.immutable()) ? DataLocation::Memory : DataLocation::Storage;
```
Comment (Marenz) [inline libsolidity/analysis/ReferencesResolver.cpp:399]: Why does `_variable.immutable()` makes the type location `Memory`?

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -480,6 +480,10 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 				"Initial value for constant variable has to be compile-time constant."
 			);
 	}
+	else if (_variable.immutable())
+		if (_variable.type()->isValueType())
```
Comment (Marenz) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
	else if (_variable.immutable() && _variable.type()->isValueType())
```

File: libsolidity/ast/AST.cpp [hunk 3]
```diff
@@ -508,7 +508,7 @@ set<VariableDeclaration::Location> VariableDeclaration::allowedDataLocations() c
 
 	if (!hasReferenceOrMappingType() || isStateVariable() || isEventParameter())
 		return set<Location>{ Location::Unspecified };
-	else if (isStateVariable() && isConstant())
+	else if (isStateVariable() && (isConstant() || immutable()))
```
Comment (Marenz) [inline libsolidity/ast/AST.cpp]: Would this `else if` ever be taken for `isStateVariable() == true`? I think the one in line 509 catches all those cases and this one will never match?

File: libsolidity/analysis/ReferencesResolver.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/ReferencesResolver.cpp:399]: If we at some point implement immutable arrays, then they will either be in memory, or we add the data location "code", but currently, this is the closest match.

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: I did it like this to keep the symmetry with the `isConstant()` check above.

File: libsolidity/ast/AST.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/ast/AST.cpp]: Right, it seems this can be removed.

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (ekpyron) [inline libsolidity/analysis/TypeChecker.cpp]: The condition is inverted, though :-):
```suggestion
		if (!_variable.type()->isValueType())
```

**Review by ekpyron — APPROVED** [review]
Not sure which of the immutable PRs should get a changelog entry :-).

**Pull request merged**